backend/src/main/java/fk/prof/backend/aggregator/AggregationWindow.java

```diff
@@ -26,93 +30,62 @@ public AggregationWindow(String appId, String clusterId, String procId,
     this.start = start;
     this.endWithTolerance = this.start.plusMinutes(windowDurationInMinutes).plusSeconds(toleranceInSeconds);
     for (int i = 0; i < workIds.length; i++) {
-      this.workStatusLookup.put(workIds[i], new ProfileWorkInfo());
+      this.workInfoLookup.put(workIds[i], new ProfileWorkInfo());
     }
   }
 
-  public String getAppId() {
-    return appId;
-  }
-
-  public String getClusterId() {
-    return clusterId;
-  }
-
-  public String getProcId() {
-    return procId;
-  }
-
-  public ProfileWorkInfo getWorkInfo(long workId) {
-    return workStatusLookup.get(workId);
-  }
+  public AggregationState startProfile(long workId, int recorderVersion, LocalDateTime startedAt) throws AggregationFailure {
+    ensureEntityIsWriteable();
 
-  public CpuSamplingAggregationBucket getCpuSamplingAggregationBucket() {
-    return this.cpuSamplingAggregationBucket;
-  }
-
-  public boolean startProfile(long workId, LocalDateTime startedAt) {
-    boolean updated = true;
-    ProfileWorkInfo workInfo = this.workStatusLookup.get(workId);
-    if (workInfo.getStatus().equals(AggregationStatus.SCHEDULED)) {
-      workInfo.setStatus(AggregationStatus.ONGOING);
-    } else if (workInfo.getStatus().equals(AggregationStatus.PARTIAL)) {
-      workInfo.setStatus(AggregationStatus.ONGOING_PARTIAL);
-    } else {
-      updated = false;
-    }
-
-    if (updated) {
-      workInfo.setStartedAt(startedAt);
+    try {
+      ProfileWorkInfo workInfo = this.workInfoLookup.get(workId);
+      return workInfo.startProfile(recorderVersion, startedAt);
+    } catch (IllegalStateException ex) {
+      throw new AggregationFailure(String.format("Error starting profile for work_id=%d, recorder_version=%d, startedAt=%s",
+          workId, recorderVersion, startedAt.toString()), ex);
     }
-
-    return updated;
   }
 
-  public boolean completeProfile(long workId) {
-    boolean updated = true;
-    ProfileWorkInfo workInfo = this.workStatusLookup.get(workId);
-    if (workInfo.getStatus().equals(AggregationStatus.ONGOING)) {
-      workInfo.setStatus(AggregationStatus.COMPLETED);
-    } else if (workInfo.getStatus().equals(AggregationStatus.ONGOING_PARTIAL)) {
-      workInfo.setStatus(AggregationStatus.RETRIED);
-    } else {
-      updated = false;
-    }
+  public AggregationState completeProfile(long workId) throws AggregationFailure {
+    ensureEntityIsWriteable();
```

Comment:
I don't think this is worth the cost of acquiring locks. I would lean towards the former where we are giving weak semantics only and can be documented explicitly. Another approach is to chuck finalize semantics altogehter and AggregationWindow can directly implement SerializableAggregationWindow (instead of FinalizedAggregationWindow). Of course, the consumer of this entity will have to be cautious about not modifying this once serialization has started (easier in our case since serialization is always initiated once aggregation window has expired)